File: services/mcp-fatsecret/src/server/tools/search_food.py
"""
MCP Tool: search_food
Searches for foods in FatSecret database.
"""

from typing import Dict, Any
import json
import logging

from ...client.fatsecret_client import FatSecretClient

logger = logging.getLogger(__name__)


async def search_food_tool(
    query: str,
    max_results: int = 10,
    page: int = 0,
    client: FatSecretClient = None
) -> str:
    """
    Search for foods in FatSecret database.
    
    Args:
        query: Food name to search for
        max_results: Maximum number of results (default: 10)
        page: Page number for pagination (default: 0)
        client: FatSecret API client instance
        
    Returns:
        JSON string with search results
    """
    if not client:
        client = FatSecretClient()
    
    try:
        foods = await client.search_foods(query, max_results, page)
        
        result = {
            "success": True,
            "query": query,
            "total_results": len(foods),
            "page": page,
            "foods": [
                {
                    "food_id": food.food_id,
                    "food_name": food.food_name,
                    "food_type": food.food_type,
                    "brand_name": food.brand_name,
                    "food_description": food.food_description
                }
                for food in foods
            ]
        }
        
        return json.dumps(result, ensure_ascii=False)
        
    except Exception as e:
        logger.error(f"Error in search_food_tool: {e}")
        return json.dumps({
            "success": False,
            "error": str(e),
            "query": query
        })


# Tool metadata for MCP
TOOL_METADATA = {
    "name": "search_food",
    "description": "Search for foods in the FatSecret database by name or description",
    "input_schema": {
        "type": "object",
        "properties": {
            "query": {
                "type": "string",
                "description": "Food name or search query (e.g., 'eggs', 'chicken breast', 'apple')"
            },
            "max_results": {
                "type": "integer",
                "description": "Maximum number of results to return",
                "default": 10
            },
            "page": {
                "type": "integer",
                "description": "Page number for pagination (starts from 0)",
                "default": 0
            }
        },
        "required": ["query"]
    }
}

File: aifood-plugin/src/tools/search-food.ts
/**
 * Search Food Tool
 * Search FatSecret database for food items
 */

import type { FatSecretClient } from '../services/fatsecret.js';
import type { SearchFoodParams } from '../types/index.js';

export function createSearchFoodTool(fatsecret: FatSecretClient) {
  return {
    name: 'search_food',
    description: 'Search the FatSecret database for food items and their nutritional information. Use this when the user wants to find calories or nutrition info for a specific food without logging it.',
    parameters: {
      type: 'object',
      properties: {
        query: {
          type: 'string',
          description: 'Food name to search for (e.g., "chicken breast", "apple", "гречка")',
        },
        maxResults: {
          type: 'number',
          description: 'Maximum number of results to return (default: 5, max: 10)',
        },
      },
      required: ['query'],
    },
    handler: async (params: SearchFoodParams) => {
      const { query, maxResults = 5 } = params;

      // Search foods
      const foods = await fatsecret.searchFoods(query, Math.min(maxResults, 10));

      if (foods.length === 0) {
        return {
          success: false,
          message: `No foods found for "${query}". Try a different search term.`,
          results: [],
        };
      }

      // Get detailed info for each food
      const results = await Promise.all(
        foods.slice(0, maxResults).map(async (food) => {
          const details = await fatsecret.getFood(food.foodId);

          if (!details || details.servings.length === 0) {
            return {
              name: food.foodName,
              brand: food.brandName,
              description: food.foodDescription,
              servings: [],
            };
          }

          // Format servings info
          const servings = details.servings.slice(0, 3).map((s) => ({
            description: s.servingDescription,
            calories: s.calories ? Math.round(s.calories) : null,
            protein: s.protein ? Math.round(s.protein * 10) / 10 : null,
            carbs: s.carbohydrate ? Math.round(s.carbohydrate * 10) / 10 : null,
            fat: s.fat ? Math.round(s.fat * 10) / 10 : null,
          }));

          return {
            name: details.foodName,
            brand: details.brandName,
            servings,
          };
        })
      );

      return {
        success: true,
        message: `Found ${results.length} results for "${query}"`,
        results,
      };
    },
  };
}
